core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java

```diff
@@ -122,86 +89,101 @@ protected void doResolve(InetSocketAddress unresolvedAddress, Promise<InetSocket
             promise.tryFailure(new IllegalStateException("resolver is closed already."));
             return;
         }
-        final CompletableFuture<CacheEntry> entryFuture =
-                cache.get(unresolvedAddress.getHostString(), s -> loadingCache(unresolvedAddress, promise));
-        assert entryFuture != null; // loader does not return null.
-        entryFuture.handle((entry, unused) -> {
-            final Throwable cause = entry.cause();
-            if (cause != null) {
-                promise.tryFailure(cause);
-                return null;
+        final String hostname = unresolvedAddress.getHostString();
+        final int port = unresolvedAddress.getPort();
+        final CacheEntry entry = addressResolverCache.getIfPresent(hostname);
+        if (entry != null) {
+            complete(promise, entry, port);
+            return;
+        }
+
+        // Duplicate queries will be blocked by CachingDnsResolver.
```

Comment:
blocked -> merged into the previous one ?

Reply:
Sounds better!